drv/ignition-api/src/lib.rs

```diff
@@ -0,0 +1,201 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! API crate for the Ignition server.
+
+#![no_std]
+
+use bitfield::bitfield;
+use derive_idol_err::IdolError;
+use drv_fpga_api::FpgaError;
+use idol_runtime::ServerDeath;
+use userlib::{sys_send, FromPrimitive, ToPrimitive};
+use zerocopy::{AsBytes, FromBytes};
+
+#[derive(
+    Copy, Clone, Debug, PartialEq, FromPrimitive, ToPrimitive, IdolError,
+)]
+pub enum IgnitionError {
+    ServerDied,
+    FpgaError,
+    InvalidPort,
+    InvalidValue,
+    NoTargetPresent,
+    RequestInProgress,
+}
+
+impl From<ServerDeath> for IgnitionError {
+    fn from(_e: ServerDeath) -> Self {
+        Self::ServerDied
+    }
+}
+
+impl From<FpgaError> for IgnitionError {
+    fn from(_e: FpgaError) -> Self {
+        Self::FpgaError
+    }
+}
+
+bitfield! {
+    #[derive(Copy, Clone, Debug, PartialEq, FromPrimitive, ToPrimitive, FromBytes, AsBytes)]
+    #[repr(C)]
+    pub struct PortState(u64);
+    pub target_present, _: 0;
+}
+
+impl PortState {
```

Comment:
You can move some of this block into the `bitfield!` type, e.g.
```rust
bitfield! {
    // ...
    pub u8, raw_receiver_status, _: 15, 8;
}
```

Alternatively, if you implement `impl From<u8> for ReceiverStatus`, you could do _everything_ in the `bitfield!` macro, removing `pub fn receiver_status` entirely:
```rust
bitfield! {
    // ...
    pub u8, into ReceiverStatus, raw_receiver_status, _: 15, 8;
}
```

You can't do exactly the same thing for `target`, since it's conditional on bit 0, but maybe adding `raw_target, _: 63, 16` then calling `raw_target()` in `pub fn target`?

(Note that this would add a shift; you'd need to adjust bit offsets in `struct Target` below)

Reply:
Looking at this a little more, I'm not sure why we're using `FromPrimitive` and `ToPrimitive` for newtypes in this file **in general**.  I'd suggest removing them whenever possible, and instead using `derive_more::{From, Into}`, i.e.

```rust
use derive_more::{From, Into};
// ...
bitfield! {
    #[derive(Copy, Clone, Debug, PartialEq, From, Into)]
    #[repr(C)]
    pub struct ReceiverStatus(u8);
    pub aligned, _: 0;
    pub locked, _: 1;
    pub polarity_inverted, _: 2;
}
```

That means that you'd be able to use the `pub u8, into ReceiverStatus, receiver_status, _: 15, 8;` form and remove a bunch of hand-written functions.

Reply:
This cleans up nicely, thanks!